src/webgpu/api/validation/copyBufferToBuffer.spec.ts

```diff
@@ -0,0 +1,263 @@
+    srcOffset,
+    dstBuffer: errorBuffer,
+    dstOffset,
+    copySize,
+    isSuccess: false,
+  });
+});
+
+g.test('buffer_usage')
+  .params([
+    { srcUsage: GPUBufferUsage.COPY_SRC, dstUsage: GPUBufferUsage.COPY_DST, isSuccess: true },
+    { srcUsage: GPUBufferUsage.COPY_DST, dstUsage: GPUBufferUsage.COPY_SRC, isSuccess: false },
+    { srcUsage: GPUBufferUsage.COPY_SRC, dstUsage: GPUBufferUsage.COPY_SRC, isSuccess: false },
+  ] as const)
+  .fn(async t => {
+    const { srcUsage, dstUsage, isSuccess } = t.params;
+
+    const bufferSize = 16;
+    const srcOffset = 0;
+    const dstOffset = 0;
+    const copySize = 8;
+
+    const srcBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: srcUsage,
+    });
+    const dstBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: dstUsage,
+    });
+
+    t.TestCopyBufferToBuffer({
+      srcBuffer,
+      srcOffset,
+      dstBuffer,
+      dstOffset,
+      copySize,
+      isSuccess,
+    });
+  });
+
+g.test('copy_size')
+  .params([
+    { copySize: 0, isSuccess: true },
+    { copySize: 2, isSuccess: false },
+    { copySize: 4, isSuccess: true },
+    { copySize: 5, isSuccess: false },
+    { copySize: 8, isSuccess: true },
+  ] as const)
+  .fn(async t => {
+    const { copySize, isSuccess } = t.params;
+
+    const bufferSize = 16;
+    const srcOffset = 0;
+    const dstOffset = 0;
+
+    const srcBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC,
+    });
+    const dstBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_DST,
+    });
+
+    t.TestCopyBufferToBuffer({
+      srcBuffer,
+      srcOffset,
+      dstBuffer,
+      dstOffset,
+      copySize,
+      isSuccess,
+    });
+  });
+
+g.test('copy_offset')
+  .params([
+    { srcOffset: 0, dstOffset: 0, isSuccess: true },
+    { srcOffset: 2, dstOffset: 0, isSuccess: false },
+    { srcOffset: 4, dstOffset: 0, isSuccess: true },
+    { srcOffset: 5, dstOffset: 0, isSuccess: false },
+    { srcOffset: 8, dstOffset: 0, isSuccess: true },
+    { srcOffset: 0, dstOffset: 2, isSuccess: false },
+    { srcOffset: 0, dstOffset: 4, isSuccess: true },
+    { srcOffset: 0, dstOffset: 5, isSuccess: false },
+    { srcOffset: 0, dstOffset: 8, isSuccess: true },
+    { srcOffset: 4, dstOffset: 8, isSuccess: true },
+  ] as const)
+  .fn(async t => {
+    const { srcOffset, dstOffset, isSuccess } = t.params;
+
+    const bufferSize = 16;
+    const copySize = 8;
+
+    const srcBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC,
+    });
+    const dstBuffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_DST,
+    });
+
+    t.TestCopyBufferToBuffer({
+      srcBuffer,
+      srcOffset,
+      dstBuffer,
+      dstOffset,
+      copySize,
+      isSuccess,
+    });
+  });
+
+g.test('copy_overflow')
+  .params([
```

Comment:
Depending on whether IDL allows it, we should also test larger numbers here (like kFirstNumberBefore2_64) - see thread: https://github.com/gpuweb/cts/pull/255#discussion_r467330003

Reply:
Sorry I am not quite clear about this. Do you mean we should use 2 ** 31 in this test like what we are doing in the test  "offset_plus_required_bytes_in_copy_overflow" ?

Reply:
I mean, numbers larger than MAX_ALIGNED_SAFE_INTEGER (which is ~`2**53`). You can copy a definition of kFirstNumberBefore2_64 from that comment and use it here. This lets us have cases where offset+size > `2**64`.

Reply:
Okay, @ponitka found that that's actually not valid (see that comment thread). So closing this thread.